### ISSUE TITLE
Keep rule order as a flag?

### ISSUE THREAD

**Issue by gabrielhdt** [Deducteam/lambdapi#331]
Hi,
Would it be sensible to use `keep-rule-order` as a flag (with `set flag ...`) rather than a command line option?

In my case, it would be to create a list membership function, so I want to write
```
symbol Utype: TYPE
symbol type_list: TYPE
symbol cons: ..., symbol nil: ...

set declared "𝔅"
constant symbol 𝔅: TYPE
constant symbol t: 𝔅
constant symbol f: 𝔅

symbol mem: Utype ⇒ type_list ⇒ 𝔅
rule mem &X (cons &X _) → t
rule mem &X (cons _ &L) → mem &X &L
rule mem _ nil                  → f
```

So first, it seems I need to keep rule order, or can I do without? Otherwise, the flag has several advantages:
+ the LSP is informed of the strategy,
+ the strategy can be changed in the file,
+ fewer arguments in command line.

**Comment by rlepigre** [comment]
I think we could do that, let's see what everyone else thinks.

By the way, the rewriting system you are defining is not confluent, so I am wondering if this is the way to go to define the `mem` function. I cannot think of any other way, so maybe this indicates that we need a kind of guard on the second rule to say that `&X` and the wildcard should not be convertible. So something like the following?
```
rule mem  &X (cons &Y &L) when not &X ≡ &Y→ mem &X &L
```
What do you think @fblanqui? Is that necessary?

**Comment by francoisthire** [comment]
> a kind of guard on the second rule to say that &X and the wildcard should not be convertible. 

I don't think this is not decidable and in particular such criterion would not be stable by substitution. The only true way to define a `mem` function is to have an decidable equality on the elements of the container. However, I think that `lambdapi` allows such shortcut is nice for the user.

**Comment by francoisthire** [comment]
There is an easy feature to implement I think is to have a flag and it can be used only once per definable symbol. It fixes many problems that could arise with non-confluent rewrite system.

> the strategy can be changed in the file,

This seems harder to implement and one need to ensure that it is not problematic. Changing the strategy while defining a recursive function could lead maybe to an inconsistency for exemple.


```
rule mem  &X (cons &Y &L) when not &X ≡ &Y → mem &X &L
```

See this PR on Dedukti : https://github.com/Deducteam/Dedukti/pull/74

Raphaël's remark made think that was a bad idea.

**Comment by fblanqui** [comment]
After the last remarks, it seems safer to remove the file-wide option --keep-rule-order and instead use a symbol-wide command, for instance `set keep_rule_order <symbol>`. This command will make LP (re)compute the decision tree of the symbol by taking into account the order of rules. But what is the order of rules when they are spread over several files? For instance, what is the order in the following situation,
```
// file A.lp
symbol A:TYPE
symbol a:A
symbol f : ...
rule f a --> a

// file B.lp
require open A
symbol b:A
rule f b --> b
// rule order here?

// file C.lp
require open A
symbol c:A
rule f c --> c

// file D.lp
require B
require C
// rule order here?

// file E.lp
require C
require B
// rule order here?
```
This should be clarified in the documentation.

**Comment by gabrielhdt** [comment]
Closed by #427
